Support for non-PIPE outputs

I have some code with uses `subprocess` with a file handle as the `stdout` argument. The `subprocess` module will write the output of the command to the file handle, and my code does not directly call `communicate`. When I register a command with  `fake_process` in the tests with some `stdout` content specified, the file doesn't seem to have the expected content written to it. Have I overlooked something, or is that not supported by this library?

Example:
```py
# tests/test_sp.py
import subprocess


def test_sp(fake_process):
    fake_process.register_subprocess(["echo", "hello"], stdout="hello\n")

    filename = "file.txt"
    with open(filename, "w") as handle:
        subprocess.run(["echo", "hello"], stdout=handle)

    assert open(filename).read() == "hello\n"
```

Result:
```
╰─❯ poetry run pytest
============================= test session starts ==============================
platform darwin -- Python 3.9.7, pytest-5.4.3, py-1.10.0, pluggy-0.13.1
rootdir: [...]/fake_sp_test
plugins: subprocess-1.3.0
collected 1 item                                                               

tests/test_sp.py F                                                       [100%]

=================================== FAILURES ===================================
___________________________________ test_sp ____________________________________

fake_process = <pytest_subprocess.core.FakeProcess object at 0x10ee90190>

    def test_sp(fake_process):
        fake_process.register_subprocess(["echo", "hello"], stdout="hello\n")
    
        filename = "file.txt"
        with open(filename, "w") as handle:
            subprocess.run(["echo", "hello"], stdout=handle)
    
>       assert open(filename).read() == "hello\n"
E       AssertionError: assert '' == 'hello\n'
E         - hello

tests/test_sp.py:11: AssertionError
=========================== short test summary info ============================
FAILED tests/test_sp.py::test_sp - AssertionError: assert '' == 'hello\n'
============================== 1 failed in 0.06s ===============================
```

--- a/changelog.d/bug.a83f8493.entry.yaml
+++ b/changelog.d/bug.a83f8493.entry.yaml
@@ -1,0 +1,5 @@
+message: Correctly handle file stream output.
+pr_ids:
+- '58'
+timestamp: 1635705281
+type: bug

--- a/pytest_subprocess/core.py
+++ b/pytest_subprocess/core.py
@@ -9,12 +9,14 @@ import time
 from collections import defaultdict
 from collections import deque
 from copy import deepcopy
+from functools import partial
 from typing import Any as AnyType
 from typing import Awaitable
 from typing import Callable
 from typing import DefaultDict
 from typing import Deque
 from typing import Dict
+from typing import IO
 from typing import List
 from typing import Optional
 from typing import Sequence
@@ -168,13 +170,18 @@ class FakePopen:
                 "different. Pass one or the other."
             )
 
-        if kwargs.get("stdout") == subprocess.PIPE:
+        stdout = kwargs.get("stdout")
+        if stdout == subprocess.PIPE:
             self.stdout = self._prepare_buffer(self.__stdout)
+        elif isinstance(stdout, (io.BufferedWriter, io.TextIOWrapper)):
+            self._write_to_buffer(self.__stdout, stdout)
         stderr = kwargs.get("stderr")
         if stderr == subprocess.STDOUT and self.__stderr:
             self.stdout = self._prepare_buffer(self.__stderr, self.stdout)
         elif stderr == subprocess.PIPE:
             self.stderr = self._prepare_buffer(self.__stderr)
+        elif isinstance(stderr, (io.BufferedWriter, io.TextIOWrapper)):
+            self._write_to_buffer(self.__stderr, stderr)
 
     def _prepare_buffer(
         self, input: OPTIONAL_TEXT_OR_ITERABLE, io_base: BUFFER = None,
@@ -214,6 +221,18 @@ class FakePopen:
         # similar as above - mypy has to be disabled because unions
         io_base.write(input)  # type: ignore
         return io_base
+
+    def _write_to_buffer(self, data: OPTIONAL_TEXT_OR_ITERABLE, buffer: IO) -> None:
+        data_type: Callable = (
+            # mypy doesn't seem to recognize `partial` as a function
+            partial(bytes, encoding=sys.getfilesystemencoding())  # type: ignore
+            if "b" in buffer.mode
+            else str
+        )
+        if isinstance(data, (list, tuple)):
+            buffer.writelines([data_type(line + "\n") for line in data])
+        else:
+            buffer.write(data_type(data))
 
     def _convert(self, input: Union[str, bytes]) -> Union[str, bytes]:
         if isinstance(input, bytes) and self.text_mode:

--- a/tests/test_subprocess.py
+++ b/tests/test_subprocess.py
@@ -1,5 +1,6 @@
 # -*- coding: utf-8 -*-
 import getpass
+import io
 import os
 import platform
 import signal
@@ -1034,3 +1035,36 @@ def test_signal_callback(fake_process):
     process.wait()
 
     assert process.returncode == -1
+
+
+@pytest.mark.parametrize("fake", [False, True])
+def test_non_piped_streams(tmpdir, fake_process, fake):
+    fake_process.allow_unregistered(not fake)
+    if fake:
+        fake_process.register_subprocess(
+            ["python", "-u", "example_script.py", "stderr"],
+            stdout=["Stdout line 1", "Stdout line 2"],
+            stderr=["Stderr line 1"],
+        )
+
+    stdout_path = tmpdir.join("stdout")
+    stderr_path = tmpdir.join("stderr")
+
+    with open(stdout_path, "w") as stdout, open(stderr_path, "w") as stderr:
+        process = subprocess.Popen(
+            ["python", "-u", "example_script.py", "stderr"],
+            stdout=stdout,
+            stderr=stderr,
+        )
+
+        err, out = process.communicate()
+
+    assert out is None
+    assert err is None
+
+    with open(stdout_path, "r") as stdout, open(stderr_path, "r") as stderr:
+        out = stdout.readlines()
+        err = stderr.readlines()
+
+    assert out == ["Stdout line 1\n", "Stdout line 2\n"]
+    assert err == [f"Stderr line 1\n"]

--- a/tests/test_subprocess.py
+++ b/tests/test_subprocess.py
@@ -1,6 +1,5 @@
 # -*- coding: utf-8 -*-
 import getpass
-import io
 import os
 import platform
 import signal
@@ -1070,7 +1069,7 @@ def test_non_piped_streams(tmpdir, fake_process, fake, bytes):
         err = stderr.readlines()
 
     assert out == ["Stdout line 1\n", "Stdout line 2\n"]
-    assert err == [f"Stderr line 1\n"]
+    assert err == ["Stderr line 1\n"]
 
 
 @pytest.mark.parametrize("fake", [False, True])

--- a/tests/test_subprocess.py
+++ b/tests/test_subprocess.py
@@ -435,7 +435,7 @@ def test_multiple_wait(fake_process, fake):
         process.wait(timeout=0.2)
 
     with pytest.raises(subprocess.TimeoutExpired):
-        process.wait(timeout=0.2)
+        process.wait(timeout=0.1)
 
     process.wait(0.4)
 

--- a/tests/test_subprocess.py
+++ b/tests/test_subprocess.py
@@ -1038,7 +1038,8 @@ def test_signal_callback(fake_process):
 
 
 @pytest.mark.parametrize("fake", [False, True])
-def test_non_piped_streams(tmpdir, fake_process, fake):
+@pytest.mark.parametrize("bytes", [True, False])
+def test_non_piped_streams(tmpdir, fake_process, fake, bytes):
     fake_process.allow_unregistered(not fake)
     if fake:
         fake_process.register_subprocess(
@@ -1050,7 +1051,9 @@ def test_non_piped_streams(tmpdir, fake_process, fake):
     stdout_path = tmpdir.join("stdout")
     stderr_path = tmpdir.join("stderr")
 
-    with open(stdout_path, "w") as stdout, open(stderr_path, "w") as stderr:
+    mode = "wb" if bytes else "w"
+
+    with open(stdout_path, mode) as stdout, open(stderr_path, mode) as stderr:
         process = subprocess.Popen(
             ["python", "-u", "example_script.py", "stderr"],
             stdout=stdout,
@@ -1068,3 +1071,36 @@ def test_non_piped_streams(tmpdir, fake_process, fake):
 
     assert out == ["Stdout line 1\n", "Stdout line 2\n"]
     assert err == [f"Stderr line 1\n"]
+
+
+@pytest.mark.parametrize("fake", [False, True])
+@pytest.mark.parametrize("bytes", [True, False])
+def test_non_piped_same_file(tmpdir, fake_process, fake, bytes):
+    fake_process.allow_unregistered(not fake)
+    if fake:
+        fake_process.register_subprocess(
+            ["python", "-u", "example_script.py", "stderr"],
+            stdout=["Stdout line 1", "Stdout line 2"],
+            stderr="Stderr line 1\n",
+        )
+
+    output_path = tmpdir.join("output")
+
+    mode = "wb" if bytes else "w"
+
+    with open(output_path, mode) as out_file:
+        process = subprocess.Popen(
+            ["python", "-u", "example_script.py", "stderr"],
+            stdout=out_file,
+            stderr=out_file,
+        )
+
+        err, out = process.communicate()
+
+    assert out is None
+    assert err is None
+
+    with open(output_path, "r") as out_file:
+        output = out_file.readlines()
+
+    assert output == ["Stdout line 1\n", "Stdout line 2\n", "Stderr line 1\n"]

